syntax = "proto3";

/*
gRPC Caption Generation Service (faster-whisper backend)

Streaming video upload in chunks:
rpc Submit (stream UploadChunk) returns (SubmitReply)

After a successful upload, the service queues the task
and processes audio â†’ segments -> WebVTT.

Statuses:
* queued      (file has been accepted and is waiting for a worker.)
* processing  (transcription in progress)
* done        (result is ready)
* error       (error)

authorization: Bearer <TOKEN>

task = "transcribe" - captions only (orig lang)
task = "translate"  - translate to english


*/

package ytcms;

// --- UploadChunk: video streaming (each chunk is a part of a file)
message UploadChunk {
  string request_id = 1; // client ID (for correlation)
  string video_id   = 2; // The video ID from YurTube app
  string lang       = 3; // "auto" or ISO code
  string task       = 4; // "transcribe" | "translate"
  bytes  data       = 5; // a piece of binary data
  bool   last       = 6; // last chunk sign
  string filename   = 7; // optional filename (for meta)
}

// --- Reply to Submit after receiving the last chunk
message SubmitReply {
  string request_id = 1;
  string job_id     = 2;
  string status     = 3; // queued | error
  string error      = 4; // if status=error
}

// --- Request status by job_id
message JobStatusRequest {
  string job_id = 1;
}

// --- Status response
message JobStatusReply {
  string job_id   = 1;
  string status   = 2;   // queued | processing | done | error
  float  progress = 3;   // 0..1
  string task     = 4;   // transcribe/translate
  string error    = 5;   // error text on status=error
}

// --- Query result
message ResultRequest {
  string job_id = 1;
}

// --- Segment descriptions
message Segment {
  double start = 1;
  double end   = 2;
  string text  = 3;
}

// --- Full result
message ResultReply {
  string job_id        = 1;
  string detected_lang = 2;
  string vtt           = 3;       // full WebVTT
  repeated Segment segments = 4;  // duplicate VTT, but may convenient for UI
  string model         = 5;
  string device        = 6;
  string compute_type  = 7;
  double duration_sec  = 8;
  string task          = 9;
}

// --- Service
service CaptionsService {
  // Stream load video.
  rpc Submit (stream UploadChunk) returns (SubmitReply);

  // One-time status.
  rpc GetStatus (JobStatusRequest) returns (JobStatusReply);

  // Streaming status update (while the task is not completed).
  rpc StreamStatus (JobStatusRequest) returns (stream JobStatusReply);

  // Getting the result (after status=done).
  rpc GetResult (ResultRequest) returns (ResultReply);
}